File: python-playwright/src/core/base_test.py
# Created on : Mar 28, 2026, 5:45:20 PM
# Author     : Tahreem J. Naseem (www.tahreems.com / +92-333-443-8775) {TECHBYTES}


import math
from pathlib import Path

import allure
from playwright.sync_api import Locator

from src.utilities.logger import get_logger, get_test_logger


class BaseTest:
    logger = get_logger(__name__)

    def setup_test(self, page, config, request):
        self.logger.info(
            "Setting up test with provided Playwright page instance, config, and request."
        )
        self.page = page
        self.config = config
        self.request = request
        self._logger = get_test_logger(request.node.nodeid)
        self.base_url = self.config.get("base_url", "")
        self.logger.debug(f"Test setup complete. Base URL: {self.base_url}")

    def log_step(self, message: str):
        self.logger.info(f"Logging test step: {message}")
        self.logger.debug(f"Attaching test step to Allure report: {message}")
        allure.attach(
            message,
            name="Test Step",
            attachment_type=allure.attachment_type.TEXT,
        )

    def assert_text_not_none(self, text: str | None, message: str) -> bool | None:
        self.logger.info(f"Asserting text: {text}")
        assert text is not None, message

    def assert_text_different(
        self, expected_text: str | None, actual_text: str | None, message: str
    ) -> bool | None:
        self.logger.info(f"Asserting text difference: {expected_text} not in {actual_text}")
        if expected_text is not None and actual_text is not None:
            assert expected_text not in actual_text, message
        else:
            raise ValueError("'None' data type provided.")

    def assert_text_matching(
        self, expected_text: str | None, actual_text: str | None, message: str
    ) -> bool | None:
        self.logger.info(f"Asserting text matches: {expected_text} in {actual_text}")
        if expected_text is not None and actual_text is not None:
            assert expected_text in actual_text, message
        else:
            raise ValueError("'None' data type provided.")

    def assert_integer_matching(
        self, expected_value: int | None, actual_value: int | None, message: str
    ) -> bool | None:
        self.logger.info(
            f"Asserting integer matches: {expected_value} == {actual_value}"
        )
        if expected_value is not None and actual_value is not None:
            assert expected_value == actual_value, message
        else:
            raise ValueError("'None' data type provided.")

    def assert_float_matching(
        self, expected_value: float | None, actual_value: float | None, message: str
    ) -> bool | None:
        self.logger.info(f"Asserting float matches: {expected_value} == {actual_value}")
        if expected_value is not None and actual_value is not None:
            assert math.isclose(expected_value, actual_value, abs_tol=0.01)
        else:
            raise ValueError("'None' data type provided.")

    def assert_visible(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is visible.")
        is_visible = locator.is_visible()
        assert is_visible, message

    def assert_hidden(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is hidden.")
        is_hidden = locator.is_hidden()
        assert is_hidden, message

    def assert_enabled(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is enabled.")
        is_visible = locator.is_enabled()
        assert is_visible, message

    def assert_disabled(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is disabled.")
        is_hidden = not locator.is_enabled()
        assert is_hidden, message

    def assert_checked(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is checked.")
        is_checked = locator.is_checked()
        assert is_checked, message

    def assert_not_checked(self, locator: Locator, message: str) -> bool | None:
        self.logger.info(f"Asserting element: {locator} is unchecked.")
        is_not_checked = not locator.is_checked()
        assert is_not_checked, message

    def assert_size_of_list(
        self, expected_size: int, actual_size: int, message: str
    ) -> bool | None:
        self.logger.info(
            f"Asserting List expected size: {expected_size} to match actual size: {actual_size}"
        )
        assert expected_size == actual_size, message

    def assert_approx_size_of_list(
        self, expected_size: list[int], actual_size: int, message: str
    ) -> bool | None:
        self.logger.info(
            f"Asserting List expected size: {expected_size} to match actual size: {actual_size}"
        )
        assert actual_size in expected_size, message

    def assert_greater(
        self, expected_size: int, actual_size: int, message: str
    ) -> bool | None:
        self.logger.info(
            f"Asserting expected size: {expected_size} to be greater than actual size: {actual_size}"
        )
        assert actual_size > expected_size, message

    def assert_list_different(
        self, old_list: list, new_list: list, message: str
    ) -> bool | None:
        self.logger.info(f"Asserting list do not match: {old_list} != {new_list}")
        if old_list is not None and new_list is not None:
            assert old_list != new_list, message
        else:
            raise ValueError("'None' data type provided.")

    def assert_file_exists(self, path: Path, message: str) -> bool | None:
        self.logger.info(f"Asserting path: {path} exists.")
        assert path.exists(), message

File: javascript-playwright/src/core/base-test.js
import fs from "node:fs";
import { expect, test as base } from "@playwright/test";
import { loadConfig } from "../configurations/configuration-loader.js";
import { getTestLogger } from "../utilities/logger.js";
import { loadScreenshotsDir } from "../utilities/path-loader.js";

export const test = base.extend({
  config: async ({}, use) => {
    await use(loadConfig(process.env.TB_ENV || "dev"));
  },
  authenticatedPage: async ({ browser, config }, use) => {
    const context = await browser.newContext({
      httpCredentials: {
        username: config.http_username || "admin",
        password: config.http_password || "admin"
      },
      ignoreHTTPSErrors: config.ignore_https_errors,
      acceptDownloads: true,
      viewport: { width: config.viewport_width, height: config.viewport_height }
    });
    const page = await context.newPage();
    page.setDefaultTimeout(config.timeout);
    await use(page);
    await context.close();
  },
  chromiumPage: async ({ playwright, config }, use) => {
    const browser = await playwright.chromium.launch({ headless: config.headless, slowMo: config.slow_mo });
    const context = await browser.newContext({
      ignoreHTTPSErrors: config.ignore_https_errors,
      acceptDownloads: true,
      viewport: { width: config.viewport_width, height: config.viewport_height }
    });
    const page = await context.newPage();
    page.setDefaultTimeout(config.timeout);
    await use(page);
    await context.close();
    await browser.close();
  },
  firefoxPage: async ({ playwright, config }, use) => {
    const browser = await playwright.firefox.launch({ headless: config.headless, slowMo: config.slow_mo });
    const context = await browser.newContext({
      ignoreHTTPSErrors: config.ignore_https_errors,
      acceptDownloads: true,
      viewport: { width: config.viewport_width, height: config.viewport_height }
    });
    const page = await context.newPage();
    page.setDefaultTimeout(config.timeout);
    await use(page);
    await context.close();
    await browser.close();
  },
  geolocationPage: async ({ browser, config }, use) => {
    const context = await browser.newContext({
      ignoreHTTPSErrors: config.ignore_https_errors,
      permissions: ["geolocation"],
      geolocation: {
        latitude: config.geolocation_latitude,
        longitude: config.geolocation_longitude
      },
      viewport: { width: config.viewport_width, height: config.viewport_height }
    });
    const page = await context.newPage();
    page.setDefaultTimeout(config.timeout);
    await use(page);
    await context.close();
  }
});

test.afterEach(async ({ page }, testInfo) => {
  if (testInfo.status !== testInfo.expectedStatus && page) {
    fs.mkdirSync(loadScreenshotsDir(), { recursive: true });
    const path = `${loadScreenshotsDir()}/${testInfo.title.replace(/[^a-zA-Z0-9_.-]/g, "_")}_${Date.now()}.png`;
    await page.screenshot({ path, fullPage: true });
    await testInfo.attach("Failure Screenshot", { path, contentType: "image/png" });
  }
});

export { expect };

export class BaseTest {
  setupTest(page, config, testInfo) {
    this.page = page;
    this.config = config;
    this.testInfo = testInfo;
    this.logger = getTestLogger(testInfo.titlePath.join("::"));
    this.baseUrl = config.base_url;
  }

  async logStep(message) {
    await this.testInfo.attach("Test Step", { body: message, contentType: "text/plain" });
  }

  assertTextNotNone(text, message) {
    expect(text, message).not.toBeNull();
  }

  assertTextDifferent(expectedText, actualText, message) {
    expect(actualText, message).not.toContain(expectedText);
  }

  assertTextMatching(expectedText, actualText, message) {
    expect(actualText, message).toContain(expectedText);
  }

  assertIntegerMatching(expectedValue, actualValue, message) {
    expect(actualValue, message).toBe(expectedValue);
  }

  assertFloatMatching(expectedValue, actualValue, message) {
    expect(actualValue, message).toBeCloseTo(expectedValue, 2);
  }

  async assertVisible(locator, message) {
    await expect(locator, message).toBeVisible();
  }

  async assertHidden(locator, message) {
    await expect(locator, message).toBeHidden();
  }

  async assertEnabled(locator, message) {
    await expect(locator, message).toBeEnabled();
  }

  async assertDisabled(locator, message) {
    await expect(locator, message).toBeDisabled();
  }

  async assertChecked(locator, message) {
    await expect(locator, message).toBeChecked();
  }

  async assertNotChecked(locator, message) {
    await expect(locator, message).not.toBeChecked();
  }

  assertSizeOfList(expectedSize, actualSize, message) {
    expect(actualSize, message).toBe(expectedSize);
  }

  assertApproxSizeOfList(expectedSizes, actualSize, message) {
    expect(expectedSizes, message).toContain(actualSize);
  }

  assertGreater(expectedSize, actualSize, message) {
    expect(actualSize, message).toBeGreaterThan(expectedSize);
  }

  assertListDifferent(oldList, newList, message) {
    expect(newList, message).not.toEqual(oldList);
  }

  assertFileExists(filePath, message) {
    expect(fs.existsSync(filePath), message).toBeTruthy();
  }
}
